src/traces/box/plot.js

```diff
@@ -109,7 +100,14 @@ function plotBoxAndWhiskers(sel, axes, trace, t) {
         bdPos1 = t.bdPos;
     }
 
-    var paths = sel.selectAll('path.box').data(Lib.identity);
+    var fn;
+    if(trace.type === 'box' ||
+        (trace.type === 'violin' && (trace.box || {}).visible)
+    ) {
+        fn = Lib.identity;
+    }
+
+    var paths = sel.selectAll('path.box').data(fn || []);
```

Comment:
I like this pattern! But two comments about it:
- `trace.box` and `trace.meanline` are [deleted](https://github.com/plotly/plotly.js/blob/c4453da2aeca42e925e67086c464dbb0b285051e/src/traces/violin/defaults.js#L49) rather than left with `visible: false` - so can't we just say `(trace.type === 'violin' && trace.box)` etc?
- I'm a bit uneasy relying on uninitialized vars being `undefined` - seems like it makes things more fragile, for example if we use `fn` again, or wrap this block in a loop or something... I'd rather a pattern like
```js
var fn;
if(...) fn = Lib.identity;
else fn = []; // or just var fn = []; and drop the else, dunno which is better

var paths = sel.selectAll('path.box').data(fn);
```